pkg/operation/common/extensions.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/flow"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// WaitUntilExtensionCRReady waits until the given extension resource has become ready.
+func WaitUntilExtensionCRReady(
+	ctx context.Context,
+	c client.Client,
+	logger *logrus.Entry,
+	newObjFunc func() runtime.Object,
+	kind string,
+	namespace string,
+	name string,
+	interval time.Duration,
+	timeout time.Duration,
+	postReadyFunc func(runtime.Object) error,
```

Comment:
The `postReadyFunc` seems like a step that isn't necessary. What do you think about leaving out `postReadyFunc`  and instead pass the `runtime.Object` in favor of `newObjFunc`, i.e. the object is then filled by `WaitUntilExtensionCRReady`. Afterwards the caller has the correctly filled `obj` and calls any post steps by himself?

Reply:
I was doing it like this earlier, but the problem was that for the `waitUntil*` functions we always should have a new object into which we load the actual state from the server. Hence, the `newObjFunc` was added in favor of the `runtime.Object`.

Reply:
I think this is also up to the caller. It's the same for the Controller-Runtime client funcs. The `newObjFunc` isn't a strong prevent mechanism, as it can also return any `runtime.Object`.

Reply:
I faced a problem when I did it like this. Concretely, e.g. in the infrastructure reconcilation the operation annotation is added. While waiting for it to be ready it fetched it the first time while the operation annotation was still present. Subsequent fetches (where the operation annotation was not present anymore) didn't overwrite the object correctly. Only when always instantiating a new object (like it was before) it was functioning correctly.

Reply:
Sure, thanks for explaining. I didn't consider this case.